python/ray/tune/trainable/trainable.py

```diff
@@ -495,13 +502,41 @@ def save(
         # User saves checkpoint
         checkpoint_dict_or_path = self.save_checkpoint(checkpoint_dir)
 
-        if _use_storage_context() and isinstance(
-            checkpoint_dict_or_path, _TrainingResult
-        ):
-            checkpoint_result = checkpoint_dict_or_path
-            assert self._last_result
-            # Update the checkpoint result to include auto-filled metrics.
-            checkpoint_result.metrics.update(self._last_result)
+        if _use_storage_context():
+            if not isinstance(self, ray.tune.trainable.FunctionTrainable):
+                # TODO(justinvyu): [cls_trainable_support]
+                # This is to get class Trainables to work in the new persistence mode.
+                # Need to handle checkpoint_dict_or_path == path, dict, or None
+                # Also need to upload to cloud, since `train.report` never gets called.
+                if isinstance(checkpoint_dict_or_path, dict):
+                    with open(
+                        os.path.join(checkpoint_dir, _DICT_CHECKPOINT_FILE_NAME), "wb"
+                    ) as f:
+                        ray_pickle.dump(checkpoint_dict_or_path, f)
+                    # Mark this directory as a dict checkpoint to check when loading.
+                    Path(checkpoint_dir).joinpath(_DICT_CHECKPOINT_MARKER).touch()
+
+                # TODO(justinvyu): Ignoring relpaths returned by save_checkpoint for now
+                local_checkpoint = NewCheckpoint.from_directory(checkpoint_dir)
+                persisted_checkpoint = self._storage.persist_current_checkpoint(
+                    local_checkpoint
+                )
+                # The checkpoint index needs to be incremented.
+                # NOTE: This is no longer using "iteration" as the folder indexing
+                # to be consistent with fn trainables.
```

Comment:
What happens to this index if we checkpoint/restore the trial? A little concerned it will get reset back to a previous value instead of monotonically incrementing in this path.

Reply:
Hmm, good point. This checkpoint index thing is very brittle I feel.

Driver and Trainable both hold a copy of a storage context. The trainable's is initialized from the driver's version, which is how manual restore loads back the checkpoint index.

On restore for trainable FT, we don't reset the index at the moment:
- curr_checkpoint_index = 3 --> Latest checkpoint was 00002
- Restore to checkpoint 00001 (how does this happen actually?)
- Next checkpoint gets saved to 3.

Reply:
> Next checkpoint gets saved to 3.

Hm since we don't reset the index, don't you mean it (incorrectly) gets saved to 2 again?

Reply:
Oh I see, so once checkpoint 2 is reported, on restore we can set that the trainable's next checkpoint is 3.

Reply:
It would be great to explicitly test these cases here, and also for the normal path if we don't have already.

Reply:
Yeah, once checkpoint 2 is reported, both the driver and trainable are aligned that the next checkpoint to save is 3.

If any restores happen in between (ex: if we restored from an older checkpoint for some reason, our next checkpoint index is still 3 (doesn't get changed on a call to restore).

Tune won't try to restore from checkpoints other than the latest though -- is the concern that the checkpoint index should be going backward in that case?

Reply:
Yes I was originally concerned about exactly that, but seems like it shouldn't be an issue.

Reply:
Ok, added some assertions on the checkpoint idxs w/ an auto-retry restore + a manual restore in the middle.